lib/seattleflu/id3c/cli/command/etl/redcap_det_uw_retrospectives.py

```diff
@@ -275,6 +284,187 @@ def determine_questionnaire_items(record: dict) -> List[dict]:
+    specific *patient_reference* and *specimen_reference*
+    """
+    clinical_results = create_clinical_result_observation_resource(redcap_record)
+    if not clinical_results:
+        return None
+
+    diagnostic_result_references = []
+    for result in clinical_results:
+        reference = create_reference(
+            reference_type = 'Observation',
+            reference = '#' + result['id']
+        )
+        diagnostic_result_references.append(reference)
+
+    collection_datetime = redcap_record['collection_date']
+
+    diagnostic_code = create_codeable_concept(
+        system = 'http://loinc.org',
+        code = '85476-0',
+        display = 'FLUAV and FLUBV and RSV pnl NAA+probe (Upper resp)'
+    )
+
+    diagnostic_report_resource = create_diagnostic_report_resource(
+        datetime = collection_datetime,
+        diagnostic_code = diagnostic_code,
+        patient_reference  = patient_reference,
+        specimen_reference = specimen_reference,
+        result = diagnostic_result_references,
+        contained = clinical_results
+    )
+
+    return (create_resource_entry(
+        resource = diagnostic_report_resource,
+        full_url = generate_full_url_uuid()
+    ))
+
+
+def create_clinical_result_observation_resource(redcap_record: dict) -> Optional[List[dict]]:
+    """
+    Determine the clinical results based on responses in *redcap_record* and
+    create observation resources for each result following the FHIR format
+    (http://www.hl7.org/implement/standards/fhir/observation.html)
+    """
+    code_map = {
+        '1240581000000104': {
+            'system': 'http://snomed.info/sct',
+            'code': '1240581000000104',
+            'display': '2019-nCoV (novel coronavirus) detected',
+        },
+        '181000124108': {
+            'system': 'http://snomed.info/sct',
+            'code': '181000124108',
+            'display': 'Influenza A virus present',
+        },
+        '441345003': {
+            'system': 'http://snomed.info/sct',
+            'code': '441345003',
+            'display': 'Influenza B virus present',
+        },
+        '441278007': {
+            'system': 'http://snomed.info/sct',
+            'code': '441278007',
+            'display': 'Respiratory syncytial virus untyped strain present',
+        },
+        '441344004': {
+            'system': 'http://snomed.info/sct',
+            'code': '441344004',
+            'display': 'Human parainfluenza virus present',
+        },
+        '186747009': {
+            'system': 'http://snomed.info/sct',
+            'code': '186747009',
+            'display': 'Coronavirus infection',  # NOT clinical finding
+        },
+        '440925005': {
+            'system': 'http://snomed.info/sct',
+            'code': '440925005',
+            'display': 'Human rhinovirus present',
+        },
+        '440930009': {
+            'system': 'http://snomed.info/sct',
+            'code': '440930009',
```

Comment:
Having snomed codes in both the key and value in this dict is a bit odd to me, but it was the easiest way to de-duplicate clinical test results without creating some ad-hoc, arbitrary intermediary name like `Flu A` to code against.

Reply:
Yep, this makes sense to me. :+1: